crates/red_knot_python_semantic/src/types.rs

```diff
@@ -74,51 +74,54 @@ fn symbol_by_id<'db>(db: &'db dyn Db, scope: ScopeId<'db>, symbol: ScopedSymbolI
 
     // If the symbol is declared, the public type is based on declarations; otherwise, it's based
     // on inference from bindings.
-    if use_def.has_public_declarations(symbol) {
-        let declarations = use_def.public_declarations(symbol);
-        // If the symbol is undeclared in some paths, include the inferred type in the public type.
-        let undeclared_ty = if declarations.may_be_undeclared() {
-            Some(
-                bindings_ty(db, use_def.public_bindings(symbol))
-                    .map(|bindings_ty| Symbol::Type(bindings_ty, use_def.public_boundness(symbol)))
-                    .unwrap_or(Symbol::Unbound),
-            )
-        } else {
-            None
-        };
-        // Intentionally ignore conflicting declared types; that's not our problem, it's the
-        // problem of the module we are importing from.
-
-        // TODO: Our handling of boundness currently only depends on bindings, and ignores
-        // declarations. This is inconsistent, since we only look at bindings if the symbol
-        // may be undeclared. Consider the following example:
-        // ```py
-        // x: int
-        //
-        // if flag:
-        //     y: int
-        // else
-        //     y = 3
-        // ```
-        // If we import from this module, we will currently report `x` as a definitely-bound
-        // symbol (even though it has no bindings at all!) but report `y` as possibly-unbound
-        // (even though every path has either a binding or a declaration for it.)
-
-        match undeclared_ty {
-            Some(Symbol::Type(ty, boundness)) => Symbol::Type(
-                declarations_ty(db, declarations, Some(ty)).unwrap_or_else(|(ty, _)| ty),
-                boundness,
-            ),
-            None | Some(Symbol::Unbound) => Symbol::Type(
-                declarations_ty(db, declarations, None).unwrap_or_else(|(ty, _)| ty),
-                Boundness::Bound,
-            ),
+    // let declaredness = use_def.public_declarations(symbol).declaredness(db);
+
+    let declarations = use_def.public_declarations(symbol);
+    let declared_ty = declarations_ty(db, declarations);
+
+    let bindings = use_def.public_bindings(symbol); // TODO: short circuit if declaredness is Bound
+    let inferred_ty = bindings_ty(db, bindings);
+
+    match declared_ty {
+        // Symbol is declared, trust the declared type
+        Ok(symbol @ Symbol::Type(_, Boundness::Bound)) => symbol,
+        // Symbol is possibly declared
+        Ok(Symbol::Type(declared_ty, Boundness::PossiblyUnbound)) => {
+            match inferred_ty {
+                // Symbol is possibly undeclared and definitely unbound
+                Symbol::Unbound => Symbol::Type(declared_ty, Boundness::Bound),
+                // Symbol is possibly undeclared and possibly bound
+                inferred @ Symbol::Type(_, Boundness::Bound) => inferred,
```

Comment:
I thought the previous behavior would have been to union declared and inferred type in this case as well as the below case? That seems like what we should do. Do we not have any test coverage of this case, or am I wrong about what the previous behavior did?

Reply:
It seems we don't have any test coverage for this. I'll note it down as a follow-up task. Changed to return a union in both cases for now.